panther_gazebo/src/led_strip.cpp

```diff
@@ -0,0 +1,279 @@
+  const gz::sim::Entity & entity, const std::shared_ptr<const sdf::Element> & sdf,
+  gz::sim::EntityComponentManager & ecm, gz::sim::EventManager & /*eventMgr*/)
+{
+  const auto model = gz::gazebo::Model(entity);
+  if (!model.Valid(ecm)) {
+    throw std::runtime_error(
+      "Error: Failed to initialize because [" + model.Name(ecm) +
+      "] (Entity=" + std::to_string(entity) +
+      ") is not a model. Please make sure that LEDStrip is attached to a valid model.");
+  }
+
+  ParseParameters(sdf);
+  light_cmd_ = SetupLightCmd(ecm);
+
+  node_.Subscribe(ns_ + "/" + image_topic_, &LEDStrip::ImageCallback, this);
+}
+
+void LEDStrip::PreUpdate(const gz::sim::UpdateInfo & info, gz::sim::EntityComponentManager & ecm)
+{
+  auto current_time = info.simTime;
+  auto period = std::chrono::milliseconds(static_cast<int>(1000 / frequency_));
+
+  if (new_image_available_ && current_time - last_update_time_ >= period) {
+    gz::msgs::Image image;
+    {
+      std::lock_guard<std::mutex> lock(image_mutex_);
+      image = last_image_;
+      new_image_available_ = false;
+    }
+
+    VisualizeLights(ecm, image);
+
+    auto light_pose = ecm.Component<gz::sim::components::Pose>(light_entity_)->Data();
+    VisualizeMarkers(image, light_pose);
+
+    last_update_time_ = current_time;
+  }
+}
+
+void LEDStrip::ParseParameters(const std::shared_ptr<const sdf::Element> & sdf)
+{
+  if (!sdf->HasElement("light_name")) {
+    ignerr << "Error: The light_name parameter is missing." << std::endl;
+  }
+  light_name_ = sdf->Get<std::string>("light_name");
+
+  if (!sdf->HasElement("topic")) {
+    ignerr << "Error: The topic parameter is missing." << std::endl;
+  }
+  image_topic_ = sdf->Get<std::string>("topic");
+
+  ns_ = sdf->HasElement("namespace") ? sdf->Get<std::string>("namespace") : ns_;
+  frequency_ = sdf->HasElement("frequency") ? sdf->Get<double>("frequency") : frequency_;
+  marker_width_ = sdf->HasElement("width") ? sdf->Get<double>("width") : marker_width_;
+  marker_height_ = sdf->HasElement("height") ? sdf->Get<double>("height") : marker_height_;
+}
+
+gz::msgs::Light LEDStrip::SetupLightCmd(gz::sim::EntityComponentManager & ecm)
+{
+  gz::msgs::Light light_cmd;
+
+  ecm.Each<gz::sim::components::Name, gz::sim::components::Light>(
+    [&](
+      const gz::sim::Entity & entity, const gz::sim::components::Name * name,
+      const gz::sim::components::Light * light_component) -> bool {
+      if (name->Data() == light_name_) {
+        light_entity_ = entity;
+        igndbg << "Light entity found: " << light_entity_ << std::endl;
+        light_cmd = ConvertLight(light_component->Data());
+        return false;  // Stop searching
+      }
+      return true;  // Continue searching
+    });
+
+  if (light_entity_ == gz::sim::kNullEntity) {
+    ignerr << "Error: Light entity not found. Return default light command msg." << std::endl;
+  }
+  return light_cmd;
+}
+
+gz::msgs::Light LEDStrip::ConvertLight(const sdf::Light & light_sdf)
```

Comment:
Could we name this function  more precisely? `ConvertLightFromSdfToMsg` `CreateLightMsgFromSdf`?